src/expr/src/interpret.rs

```diff
@@ -0,0 +1,1180 @@
+        ExtractTimestamp(funcs::ExtractTimestamp(units))
+        | ExtractTimestampTz(funcs::ExtractTimestampTz(units)) => match units {
+            DateTimeUnits::Epoch
+            | DateTimeUnits::Millennium
+            | DateTimeUnits::Century
+            | DateTimeUnits::Decade
+            | DateTimeUnits::Year => Yes,
+            _ => No,
+        },
+        DateTruncTimestamp(_) | DateTruncTimestampTz(_) => Yes,
+        // Negation is monotonically decreasing, but that's fine.
+        NegInt64(_) | NegNumeric(_) | Not(_) => Yes,
+        // Trivially monotonic, since all non-null values map to `false`.
+        IsNull(_) => Yes,
+        AbsInt64(_) => No,
+        _ => Maybe,
+    }
+}
+
+fn binary_monotonic(func: &BinaryFunc) -> (Monotonic, Monotonic) {
+    use BinaryFunc::*;
+    use Monotonic::*;
+
+    match func {
+        AddInt64 | MulInt64 | AddNumeric | MulNumeric | SubInt64 | SubNumeric => (Yes, Yes),
+        AddInterval | SubInterval => (Yes, Yes),
+        AddTimestampInterval
+        | AddTimestampTzInterval
+        | SubTimestampInterval
+        | SubTimestampTzInterval => (Yes, Yes),
+        // Monotonic in the left argument, but the right hand side has a discontinuity.
+        // Could be treated as monotonic if we also captured the valid domain of the function args.
+        DivInt64 | DivNumeric => (Yes, No),
+        Lt | Lte | Gt | Gte => (Yes, Yes),
+        _ => (Maybe, Maybe),
+    }
+}
+
+fn variadic_monotonic(func: &VariadicFunc) -> Monotonic {
+    use Monotonic::*;
+    use VariadicFunc::*;
+    match func {
+        And | Or => Yes,
+        _ => Maybe,
+    }
+}
+
+/// An inclusive range of non-null datum values.
+#[derive(Clone, Eq, PartialEq, Debug)]
+enum Values<'a> {
+    /// This range contains no values.
+    Empty,
+    /// An inclusive range. Invariant: the first element is always <= the second.
+    Within(Datum<'a>, Datum<'a>),
+    /// Constraints on structured fields, useful for recursive structures like maps.
+    /// Fields that are not present in the map default to Values::All.
+    Nested(BTreeMap<Datum<'a>, ResultSpec<'a>>),
+    /// This range might contain any value. Since we're overapproximating, this is often used
+    /// as a "safe fallback" when we can't determine the right boundaries for a range.
+    All,
+}
+
+impl<'a> Values<'a> {
+    fn just(a: Datum<'a>) -> Values<'a> {
+        match a {
+            Datum::Map(datum_map) => Values::Nested(
+                datum_map
+                    .iter()
+                    .map(|(key, val)| (key.into(), ResultSpec::value(val)))
+                    .collect(),
+            ),
+            other => Self::Within(other, other),
+        }
+    }
+
+    fn union(self, other: Values<'a>) -> Values<'a> {
+        match (self, other) {
+            (Values::Empty, r) => r,
+            (r, Values::Empty) => r,
+            (Values::Within(a0, a1), Values::Within(b0, b1)) => {
```

Comment:
We can lose a lot of precision when combining two `Values::just(...)`---e.g., `Values::just(1).union(Values::just(100))` will be `Values::within(1, 100)` even though we can really only hit the endpoints.

If these pushed-down filters end up being used for more than temporal queries (which I expect to mostly be using inequalities), it could be worthwhile to include a separate notion of finite set; we'll probably want to cap it at some low number of elements, after which we can widen to use a range.